File: alpha_assignment_mgmt/overrides/project.py
import frappe


def before_insert(doc, method):
	"""Auto-populate Project from linked Assignment Origination."""
	if doc.custom_assignment_origination:
		origination = frappe.get_doc(
			"Alpha Assignment Origination", doc.custom_assignment_origination
		)
		if origination.workflow_state != "Approved":
			frappe.throw(
				f"Cannot create Project. Assignment Origination "
				f"{origination.name} has not been approved yet."
			)

		doc.customer = origination.customer or doc.customer
		doc.custom_service_line = origination.service_line
		doc.project_type = origination.service_line
		doc.custom_branch_manager = origination.lead_branch_manager
		doc.custom_engagement_manager = origination.engagement_manager
		doc.custom_client_owner = doc.custom_client_owner or origination.client_owner
		doc.custom_risk_rating = origination.risk_rating

		if not doc.expected_end_date:
			doc.expected_end_date = (
				origination.regulatory_deadline
				or origination.statutory_deadline
			)


def validate(doc, method):
	"""Validate project fields."""
	pass


def on_update(doc, method):
	"""Create SLA for new projects, update origination, check closure."""
	old_doc = doc.get_doc_before_save()
	is_new = old_doc is None

	if is_new and doc.custom_assignment_origination and not doc.custom_engagement_sla:
		create_sla(doc)

	update_origination_status(doc, is_new)
	check_project_closure(doc)


def create_sla(doc):
	"""Auto-create Engagement SLA when Project Type is set."""
	if not doc.project_type:
		return

	sla_level_map = {
		"Tax Compliance": "SLA A - Urgent Statutory",
		"TRA Support": "SLA A - Urgent Statutory",
		"Audit Readiness": "SLA B - Audit Readiness",
		"Monthly Bookkeeping": "SLA C - Monthly Bookkeeping",
		"Accounting Reconstruction": "SLA D - Reconstruction",
		"Advisory": "SLA E - Advisory/Research/Training",
		"ERPNext Implementation": "SLA E - Advisory/Research/Training",
	}

	origination = None
	if doc.custom_assignment_origination:
		origination = frappe.get_doc(
			"Alpha Assignment Origination", doc.custom_assignment_origination
		)

	deadline = doc.expected_end_date
	if origination:
		deadline = (
			origination.regulatory_deadline
			or origination.statutory_deadline
			or doc.expected_end_date
		)

	if not deadline:
		from frappe.utils import add_days
		deadline = add_days(doc.expected_start_date or frappe.utils.today(), 30)

	sla = frappe.new_doc("Alpha Engagement SLA")
	sla.project = doc.name
	sla.customer = doc.customer
	sla.assignment_origination = doc.custom_assignment_origination
	sla.sla_level = sla_level_map.get(doc.project_type, "SLA C - Monthly Bookkeeping")
	sla.engagement_manager = doc.custom_engagement_manager
	sla.branch_manager = doc.custom_branch_manager
	sla.alpha_processing_deadline = deadline
	sla.client_due_date = deadline
	sla.insert(ignore_permissions=True)
	sla.submit()

	doc.db_set("custom_engagement_sla", sla.name)

	# Back-link SLA to Origination
	if doc.custom_assignment_origination:
		frappe.db.set_value(
			"Alpha Assignment Origination",
			doc.custom_assignment_origination,
			"sla_reference",
			sla.name,
		)


def update_origination_status(doc, is_new):
	"""Update Assignment Origination when Project is created."""
	if is_new and doc.custom_assignment_origination:
		frappe.db.set_value("Alpha Assignment Origination", doc.custom_assignment_origination, {
			"project_created": 1,
			"project_reference": doc.name,
			"workflow_state": "Project Created",
			"acceptance_status": "Approved",
			"closure_status": "In Progress",
		})


def check_project_closure(doc):
	"""Prevent closing Project without Closure Certificate."""
	if doc.status == "Completed" and not doc.custom_closure_certificate:
		frappe.throw(
			"Cannot close Project. Assignment Closure Certificate is required."
		)

File: alpha_assignment_mgmt/public/js/project.js
// Copyright (c) 2026, Alpha Associates (T) Limited and contributors
// For license information, please see license.txt

frappe.ui.form.on("Project", {
	refresh(frm) {
		if (frm.doc.custom_assignment_origination) {
			frm.add_custom_button(__("Open Origination"), () => {
				frappe.set_route("Form", "Alpha Assignment Origination", frm.doc.custom_assignment_origination);
			}, __("Navigate"));
		}
		if (frm.doc.custom_engagement_sla) {
			frm.add_custom_button(__("Open SLA"), () => {
				frappe.set_route("Form", "Alpha Engagement SLA", frm.doc.custom_engagement_sla);
			}, __("Navigate"));
		}
		if (frm.doc.custom_closure_certificate) {
			frm.add_custom_button(__("Open Closure Certificate"), () => {
				frappe.set_route("Form", "Assignment Closure Certificate", frm.doc.custom_closure_certificate);
			}, __("Navigate"));
		}

		if (frm.doc.status === "Active") {
			frm.add_custom_button(__("New Task"), () => {
				const task = frappe.model.get_new_doc("Task");
				task.project = frm.doc.name;
				frappe.set_route("Form", "Task", task);
			}, __("Actions"));
		}
	},

	custom_assignment_origination(frm) {
		if (!frm.doc.custom_assignment_origination) {
			return;
		}
		frappe.call({
			method: "frappe.client.get_value",
			args: {
				doctype: "Alpha Assignment Origination",
				filters: { name: frm.doc.custom_assignment_origination },
			fieldname: [
				"customer",
				"service_line",
				"risk_rating",
				"engagement_manager",
				"lead_branch_manager",
				"client_owner",
				"branch",
				"proposed_fee",
				"billing_method",
				"statutory_deadline",
				"regulatory_deadline",
				"urgency_level",
			],
			},
			callback(r) {
				if (r.message) {
					const d = r.message;
					if (d.customer) frm.set_value("customer", d.customer);
					if (d.service_line) {
						frm.set_value("custom_service_line", d.service_line);
						frm.set_value("project_type", d.service_line);
					}
					if (d.risk_rating) frm.set_value("custom_risk_rating", d.risk_rating);
					if (d.engagement_manager) frm.set_value("custom_engagement_manager", d.engagement_manager);
					if (d.lead_branch_manager) frm.set_value("custom_branch_manager", d.lead_branch_manager);
					if (d.client_owner) frm.set_value("custom_client_owner", d.client_owner);
					if (d.branch) frm.set_value("branch", d.branch);
					if (d.proposed_fee) frm.set_value("estimated_costing", d.proposed_fee);
					if (d.statutory_deadline && !frm.doc.expected_end_date) {
						frm.set_value("expected_end_date", d.statutory_deadline);
					} else if (d.regulatory_deadline && !frm.doc.expected_end_date) {
						frm.set_value("expected_end_date", d.regulatory_deadline);
					}
					frappe.show_alert({ message: __("Fields populated from Assignment Origination."), indicator: "green" });
				}
			},
		});
	},
});
